README.md
# tirbi

Tirbi is a remote cache implementation that is compatible with the
[remote caching](https://turborepo.org/docs/core-concepts/remote-caching)
feature of [turborepo](https://turborepo.org). It supports storing cached assets
on the file system, in memory, or in Google Cloud Storage bucket. Feel free to
contribute other storage backends.

Tirbi can be used either as a command line server, or as a fastify plugin.

## Quick start

Run tirbi server:

```shell
npx tirbi
```

This starts server that allows any authentication token and stores cached data
in memory.

Run a build with turborepo that uses tirbi as cache:

```shell
yarn turbo run build \
     --team=not_used --token="anything" --remote-only \
     --api=http://localhost:8080
```

The turbo build should be loading and saving cached artifacts to the tirbi
server.

Run `tirbi --help` to see the list of command line options.

The server can also be configured using environment variables:

- `HOST` - Host to bind to
- `PORT` - Port to bind to
- `STORAGE` - Storage URI
- `TOKEN` - An auth token

## fastify plugin

Tirbi exports a `tirbiPlugin` object that is a fastify plugin:

```typescript
import fastify from 'fastify';
import { tirbiPlugin } from 'tirbi';

async function main() {
  const server = fastify({ logger: true });
  await server.register(tirbiPlugin, {
    storage: { kind: 'memory' },
    tokens: ['s3cr3t'],
  });
  await server.listen(3030, '0.0.0.0');
}

main();
```

Have a look in [`cli.ts`](./src/cli.ts) to see how the tirbi CLI starts a
server.

## Reference

Interface describing the options of the fastify plugin.

### `TirbiOptions`

- `tokens` - (optional) an array of allowed authorization tokens as strings. If
  omitted, authorization tokens will not be checked.
- `storage` - an object configuring the storage backend. See `StorageOptions`

### `StorageOptions`

There are TypeScript types for the supported storage backends:

- `FileSystemStorageOptions` - Options for file system storage backend.
- `MemoryStorageOptions` - Options for in-memory storage backend.
- `GcsStorageOptions` - Options for Google Cloud Storage backend.
- `StorageOptions` - The union of the backend option types.

### `parseStorageUri`

Parse a URI string to a `StorageOptions` object. Returns null if there if the
URI is invalid, or uses an unknown scheme. This function will never throw an
error.

### `tirbiPlugin`

A fastify plugin. Takes a `TirbiOptions` argument.

## Docker usage

The following dockerfile lets you run tirbi in docker:

```dockerfile
FROM node:16-alpine3.14
RUN npm install -g tirbi
CMD ["tirbi"]
```

Use environment variables the control the settings of the server.

## Compatibility

Tirbi has been tested with turborepo versions 1.2.1-1.2.16 and 1.3.0-1.3.1

## Caveats

- Turborepo requires a `team` to be set when using remote caches. tirbi ignores
  the team setting, so it can be set to anything
- There's a bug that only lets you pass in a single auth token to tirbi when
  using environment variables.
- Turborepo sends requests to an `events` URL, with information about what it's
  been doing. Tirbi ignores this call, so it will turn up in the logs as a 404
  request.

## Misc

### Liveness probe

The cli listens for requests to `/healthz`, end replies with a 204. This is
useful if running in docker or other managed platforms that probes a service for
its liveness status.

### Pretty logs

To get nicely formatted logs on the command line, use
[`pino-pretty`](https://github.com/pinojs/pino-pretty):

```shell
tirbi | pino-pretty
```

yarn lint-staged

## Publishing

Uses
[`standard-version`](https://github.com/conventional-changelog/standard-version).

Can be invoked with `yarn release`.

- `yarn release --release-as major`
- `yarn release --release-as minor`
- `yarn release --release-as patch`
- `yarn release --prerelease`

After this is done, use the following:

```sh
git push --follow-tags origin main && npm publish --tag prerelease
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See
[standard-version](https://github.com/conventional-changelog/standard-version)
for commit guidelines.

### [1.0.1](https://github.com/runeh/tirbi/compare/v1.0.0...v1.0.1) (2022-06-28)

### Bug Fixes

- fix bug with gs URL parser
  ([9319658](https://github.com/runeh/tirbi/commit/9319658048e87aac87c5994775f521a86ca882e9))

## [1.0.0](https://github.com/runeh/tirbi/compare/v1.0.0-beta.9...v1.0.0) (2022-06-19)

## 1.0.0-beta.9 (2022-06-18)

### Features

- add liveness probe url
  ([d5f6758](https://github.com/runeh/tirbi/commit/d5f6758fcebfba2b091d44f36677ad08ea1d364d))

### Bug Fixes

- Add more docs
  ([d53b66d](https://github.com/runeh/tirbi/commit/d53b66d9b5117ae0d9818810ab9befd264ad8080))
- allow read to return promise
  ([55f813f](https://github.com/runeh/tirbi/commit/55f813fa4dcde8b07ee16bdaed5ed5275fb1c318))
- emit types and hook up entrypoints
  ([b5ef0a9](https://github.com/runeh/tirbi/commit/b5ef0a91ed5a16d422a304aa5f2d415181f30755))
- Fix broken storage URL parsing
  ([7eb7502](https://github.com/runeh/tirbi/commit/7eb7502fff7da9439c4b4c4e8a0127d5fe3ffc2b))
- Remove a console.log
  ([ff79150](https://github.com/runeh/tirbi/commit/ff791506795c5dac18c9356412889a79a6e53fdc))
- use correct scheme for fs urls
  ([60d1b11](https://github.com/runeh/tirbi/commit/60d1b110ce7b487d40507debcbabcc03af965c7d))


src/cli.ts
#!/usr/bin/env node

import { setTimeout } from 'timers/promises';
import { readFileSync } from 'fs';
import { resolve } from 'path';
import { InvalidArgumentError, Option, program } from 'commander';
import fastify, { FastifyInstance } from 'fastify';
import { StorageOptions, parseStorageUri } from './common';
import { tirbiPlugin } from './plugin';

const defaultHost = '0.0.0.0';
const defaultPort = 8080;
const defaultStorage: StorageOptions = { kind: 'memory' };
const defaultLivenessPath = '/healthz';
const version = readVersion();

function readVersion() {
  const raw = readFileSync(resolve(__dirname, '..', 'package.json'), 'utf-8');
  const parsed = JSON.parse(raw) as { version: string };
  return parsed.version;
}

function closeServer(server: FastifyInstance): Promise<true> {
  return new Promise((resolve) => server.close(() => resolve(true)));
}

async function sleep(ms: number): Promise<false> {
  await setTimeout(ms);
  return false;
}

async function shutdown(
  server: FastifyInstance,
  signal: string,
  gracePeriodMs: number,
) {
  server.log.info(
    { signal: signal, gracePeriodMs },
    `Received signal ${signal}`,
  );
  const closedInGracePeriod = await Promise.race([
    closeServer(server),
    sleep(gracePeriodMs),
  ]);
  if (!closedInGracePeriod) {
    server.log.warn('Server did not close within grace period');
  }
  server.log.info('Exiting');
  process.exit(0);
}

function parsePortArg(raw: string) {
  const port = Number(raw);
  if (Number.isNaN(port)) {
    throw new InvalidArgumentError('Port must be a number');
  }

  if (port < 1 || port > 65535) {
    throw new InvalidArgumentError('Port must be in range 1-65535');
  }
  return port;
}

function parseStorageArg(raw: string): StorageOptions {
  const parsed = parseStorageUri(raw);
  if (!parsed) {
    throw new InvalidArgumentError('Invalid storage URI');
  } else {
    return parsed;
  }
}

const exampleText = `
Example:

Run a tirbi server on port 3131, allowing calls using "s3cret" and "secr3t" as
bearer tokens and saving assets to the "turborepo" folder in a Google Storage
bucket named "ci-caches".

  $ tirbi --port 3131 --host 0.0.0.0 \\
          --token s3cret --token secr3t \\
          --storage gs://ci-caches/turborepo

Storage URIs:

The following storage URI formats are allowed:

- gs://bucket-name/folder - a Google cloud storage bucket with optional folder.
- fs:/tmp/assets - the '/tmp/assets' folder on the local disk
- s3://foo?region=us-west-1 an AWS S3 bucket named foo in region us-west-1
- memory:// - in-memory storage.
- memory://?sizeMb=256 - in-memory storage with a max storage size

Running the program with no arguments starts a tirbi server on port 8080 that
allows any bearer token and stores artifacts in memory.
`;

program
  .name('tirbi')
  .addOption(
    new Option('-p, --port <port>', 'Port to listen to')
      .argParser(parsePortArg)
      .default(defaultPort)
      .env('PORT'),
  )
  .addOption(
    new Option('-h, --host <host>', 'Host to bind to')
      .default(defaultHost)
      .env('HOST'),
  )
  .addOption(
    new Option('-t, --token <token...>', 'One or more auth tokens')
      .default(undefined, 'Accepts any token')
      .env('TOKEN'),
  )
  .addOption(
    new Option('-s, --storage <URI>', 'Storage backend')
      .default(defaultStorage, 'in-memory storage')
      .argParser(parseStorageArg)
      .env('STORAGE'),
  )
  .addOption(
    new Option('-l, --liveness <path>', 'Path for liveness check').env(
      'LIVENESS',
    ),
  )
  // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
  .version(version, '-v, --version', 'show tirbi version')
  .addHelpText('after', exampleText);

interface ParseOptions {
  port: number;
  host: string;
  storage: StorageOptions;
  token?: string[];
  liveness: string;
}

async function main() {
  program.parse();
  // Not really type safe, but good enough
  const { token, host, storage, port, liveness }: ParseOptions = program.opts();
  const tokens = token ?? [];
  const livenessPath = liveness ?? defaultLivenessPath;

  const server = fastify({ logger: true });
  await server.register(tirbiPlugin, { storage, tokens });

  server.log.info(`Adding liveness probe on "${livenessPath}"`);
  server.get(livenessPath, (_, reply) => {
    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    reply.code(204).send();
  });

  // If the process is interrupted, try to shutdown properly, with a grace
  // period of 5 seconds for requests to finish. If there are pending requests
  // after that,just shut down the process.
  for (const signal of ['SIGTERM', 'SIGINT']) {
    process.on(signal, () => {
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      shutdown(server, signal, 5000);
    });
  }

  await server.listen(port, host);
}

// eslint-disable-next-line @typescript-eslint/no-floating-promises
main();
